packages/devops/scripts/lambda/utilities.py

```diff
@@ -0,0 +1,236 @@
+
+
+def get_gateway_elb(tupaia_instance_name):
+    elbs = elbv2.describe_load_balancers(PageSize=400)
+    tags_by_elb = elbv2.describe_tags(
+        ResourceArns=list(map(lambda x: x['LoadBalancerArn'], elbs['LoadBalancers']))
+    )
+    matching_arns = list(filter(lambda x: tags_contains(x['Tags'], 'Type', 'Tupaia-Gateway') and tags_contains(x['Tags'], 'TupaiaInstanceName', tupaia_instance_name), tags_by_elb['TagDescriptions']))
+    if len(matching_arns) == 0:
+        raise Exception('No gateway elb found tagged with tupaia instance name: ' + tupaia_instance_name)
+    matching_arn = matching_arns[0]['ResourceArn']
+    return list(filter(lambda x: x['LoadBalancerArn'] == matching_arn, elbs['LoadBalancers']))[0]
+
+
+def get_gateway_target_group(tupaia_instance_name):
+    target_groups = elbv2.describe_target_groups(PageSize=400)
+    tags_by_target_group = elbv2.describe_tags(
+        ResourceArns=list(map(lambda x: x['TargetGroupArn'], target_groups['TargetGroups']))
+    )
+    matching_arns = list(filter(lambda x: tags_contains(x['Tags'], 'Type', 'Tupaia-Gateway') and tags_contains(x['Tags'], 'TupaiaInstanceName', tupaia_instance_name), tags_by_target_group['TagDescriptions']))
+    if len(matching_arns) == 0:
+        raise Exception('No gateway target group found tagged with tupaia instance name: ' + tupaia_instance_name)
+    matching_arn = matching_arns[0]['ResourceArn']
+    return list(filter(lambda x: x['TargetGroupArn'] == matching_arn, target_groups['TargetGroups']))[0]
+
+
+def get_gateway_listeners(tupaia_instance_name):
+    elb = get_gateway_elb(tupaia_instance_name)
+    return elbv2.describe_listeners(
+        LoadBalancerArn=elb['LoadBalancerArn'],
+        PageSize=100
+    )['Listeners']
+
+
+def create_gateway_elb(tupaia_instance_name, config, gateway_name):
+    response = elbv2.create_load_balancer(
+        Name=gateway_name,
+        Subnets=list(map(lambda x: x['SubnetId'], config['AvailabilityZones'])),
+        SecurityGroups=config['SecurityGroups'],
+        Scheme=config['Scheme'],
+        Tags=[
+            {
+                'Key': 'Type',
+                'Value': 'Tupaia-Gateway'
+            },
+            {
+                'Key': 'TupaiaInstanceName',
+                'Value': tupaia_instance_name
+            },
+        ],
+        Type=config['Type'],
+        IpAddressType=config['IpAddressType']
+    )
+    return response['LoadBalancers'][0]
+
+
+def create_gateway_target_group(tupaia_instance_name, config, target_group_name):
+    response = elbv2.create_target_group(
+        Name=target_group_name,
+        Protocol=config['Protocol'],
+        ProtocolVersion=config['ProtocolVersion'],
+        Port=config['Port'],
+        VpcId=config['VpcId'],
+        HealthCheckProtocol=config['HealthCheckProtocol'],
+        HealthCheckPort=config['HealthCheckPort'],
+        HealthCheckEnabled=config['HealthCheckEnabled'],
+        HealthCheckPath=config['HealthCheckPath'],
```

Comment:
In future it would be nice to have a slightly smarter health check, that actually makes sure the service is available rather than just checking the instance is up and responding via nginx. Out of scope for now, and better to design that system once we have more independent scaling going on so it's clearer what healthy means in each case

Reply:
Will keep in mind for scaling in the next couple weeks